computer_vision/haralick_descriptors.py

```diff
@@ -0,0 +1,433 @@
+    """
+    Apply binary mask, or thresholding based
+    on bit mask value (mapping mask is binary).
+
+    Returns the mapped true value mask and its complementary false value mask.
+
+    Example:
+        >>> img = np.array([[[108, 201, 72], [255, 11,  127]],
+        ...                 [[56,  56,  56], [128, 255, 107]]])
+        >>> gray = grayscale(img)
+        >>> binary = binarize(gray)
+        >>> morphological = opening_filter(binary)
+        >>> binary_mask(gray, morphological)
+        (array([[1, 1],
+               [1, 1]], dtype=uint8), array([[158,  97],
+               [ 56, 200]], dtype=uint8))
+    """
+    true_mask, false_mask = np.array(image_gray, copy=True), np.array(
+        image_gray, copy=True
+    )
+    true_mask[image_map == 1] = 1
+    false_mask[image_map == 0] = 0
+
+    return true_mask, false_mask
+
+
+def matrix_concurrency(image: np.ndarray, coordinate: list) -> np.ndarray:
+    """
+    Calculate sample co-occurrence matrix based on input image
+    as well as selected coordinates on image.
+
+    Implementation is made using basic iteration,
+    as function to be performed (np.max) is non-linear and therefore
+    not callable on the frequency domain.
+
+    Example:
+        >>> img = np.array([[[108, 201, 72], [255, 11,  127]],
+        ...                 [[56,  56,  56], [128, 255, 107]]])
+        >>> gray = grayscale(img)
+        >>> binary = binarize(gray)
+        >>> morphological = opening_filter(binary)
+        >>> mask_1 = binary_mask(gray, morphological)[0]
+        >>> matrix_concurrency(mask_1, [0, 1])
+        array([[0., 0.],
+               [0., 0.]])
+    """
+    matrix = np.zeros([np.max(image) + 1, np.max(image) + 1])
+
+    offset_x, offset_y = coordinate[0], coordinate[1]
+
+    for x in range(1, image.shape[0] - 1):
+        for y in range(1, image.shape[1] - 1):
+            base_pixel = image[x, y]
+            offset_pixel = image[x + offset_x, y + offset_y]
+
+            matrix[base_pixel, offset_pixel] += 1
+    matrix_sum = np.sum(matrix)
+    return matrix / (1 if matrix_sum == 0 else matrix_sum)
+
+
+def haralick_descriptors(matrix: np.ndarray) -> list:
+    """Calculates all 8 Haralick descriptors based on co-occurence input matrix.
+    All descriptors are as follows:
+    Maximum probability, Inverse Difference, Homogeneity, Entropy,
+    Energy, Dissimilarity, Contrast and Correlation
+
+    Args:
+        matrix: Co-occurence matrix to use as base for calculating descriptors.
+
+    Returns:
+        Reverse ordered list of resulting descriptors
+
+    Example:
+        >>> img = np.array([[[108, 201, 72], [255, 11,  127]],
+        ...                 [[56,  56,  56], [128, 255, 107]]])
+        >>> gray = grayscale(img)
+        >>> binary = binarize(gray)
+        >>> morphological = opening_filter(binary)
+        >>> mask_1 = binary_mask(gray, morphological)[0]
+        >>> concurrency = matrix_concurrency(mask_1, [0, 1])
```

Comment:
```suggestion
        >>> concurrency = matrix_concurrency(mask_1, (0, 1))
```
Change list to tuple (see previous comments)